miniml/compiler/compile.scm

```diff
@@ -1093,35 +1117,173 @@
+     ; (first column, other columns, rhs)
+     (let*
+      ((m (map (match-lambda (((p . ps) e) (list p ps e))) m))
+       (m (simplify-matrix env arg m))
+       (groups (split-matrix-in-groups env m))
+       (code (lower-match-groups env arg args groups)))
+      code
+     ))
+    ))
+
+(define (simplify-matrix env arg m)
+  (map (lambda (row) (simplify-row env arg row)) m))
+
+; rows of a "simplified (non-empty) matrix" are of the form
+; (head head-subpatterns other-patterns right-hand-side)
+;
+; a head is either #nil, for the wildcard pattern, or a "strong head"
+; representing the head constructor of the values being matched.
+(define (simplify-row env arg row)
+  (match-let (((p ps e) row))
+   (match p
+    (('PVar v)
+     (let* ((p (list 'PAs (list 'PWild) v)))
+     (simplify-row env arg (list p ps e))))
+    (('PAs p v)
+     (let* ((e (list 'LLet v (lid->evar arg) e)))
+     (simplify-row env arg (list p ps e))))
+    (('PWild)
+     (list #nil #nil ps e))
+    (('PInt n)
+     (list (list 'HPInt n) #nil ps e))
+    (('PConstr c l)
+     (let* ((arity (pconstr-arity env c l))
+            (l (adjust-pconstr-args l arity)))
+       (list (list 'HPConstr c (pconstr-arity env c l)) l ps e))))))
+
+; two things may go wrong here:
+; - the environment may not know the arity of the declaration,
+;   in which case we trust the pattern itself
+; - the pattern may use a single wildcard for several wildcards,
+;   in which case we trust the environment (if available)
+(define (pconstr-arity env c l)
+  (let* ((cdef (env-get-constr env c))
+         (env-arity (constr-get-arity cdef))
+         (l-arity (length l)))
+    (if (= env-arity -1) l-arity
+      (match l
+        ((('PWild))
+         env-arity)
+        (_
+         (assert (equal? l-arity env-arity))
+         l-arity)))))
+
+(define (omegas-for-pattern-head h)
+  (match h
+   (#nil #nil)
+   (('HPInt n) #nil)
+   (('HPConstr c arity) (make-list arity (list 'PWild)))))
+
+(define (split-matrix-in-groups env simplified-matrix)
+  (let*
+      (; first we compute all groups present in the matrix
+       (groups
+        (fold (lambda (row groups)
+                (match-let (((h hps ps e) row))
+                 (vhash-replace h #nil groups))
+               ) vlist-null simplified-matrix))
+       ; then we populate all the groups by adding rows with the corresponding head
+       (groups
+        (fold-right (lambda (row groups)
+                (match-let (((h hps ps e) row))
+                 (assert (pair? (vhash-assoc h groups)))
+                 (if (null? h)
+                   (begin
+                     (assert (null? hps))
+                     ; in the wildcard case, we add the row to the default group
+                     ; (representing any value whose head is not among the other groups)
+                     ; *and* to each group (their values also match a wildcard)
+                     ; this is why we need to have a pre-computed list of all groups
```

Comment:
Yes indeed. I think that duplicating code is problematic in principle, so I was planning to tackle this when introducing or-patterns (which also duplicate the actions of some rows). During compilation we can detect which actions (= right-hand-side expressions) get duplicated. Then the question is what code to generate for those. There are three potential approaches:
1. We could simply generate local functions and jump to them; this gives fairly bad code, but we probably don't care so much (especially if we disable deduplication when the duplicated action is a variable or a constant).
2. We could manually generate bytecode labels and use branches; the problem is that, in the or-pattern case, the duplicated branches may have bound the same variables in different places in the stack, so to do this correctly the easier way is to re-push the shared variables in some agreed-upon order before branching. This sounds too painful to me.
3. We could implement support for OCaml's `Lstaticcach/Lstaticraise`, which are block-local exceptions (with parameters), used in particular for deduplication within the pattern-matching compiler; or we could expose this as non-leaking local functions that can only be tail-called. This gives a nice/principled structure to do (2).

Reply:
Note: in [Data.value_compare](https://github.com/Ekdohibs/camlboot/blob/74bdd14/interpreter/data.ml#L281) there is actually no issue, as the "default" actions are very small (either `failwith <string literal>` or `assert false`).

Reply:
I started looking at implementing Lstaticraise, or rather something like

```
let-exits
  j1 x y z = ...
  j2 x y = ...
in
  ....  exit j2 foo bar ...
```

but I am tempted to do some refactoring before that (ideally to separate a "typing environment" from a "compilation environment", with variables in both sides, one just for the shape and the other just for the location), and I would prefer to do this on top of the current state of the branch, rather than on top of master, because some things are cleaner in the branch, notably the handling of `"_"`.

I hesitate between:
- suggesting to decide that duplication is not a concern for now, and that this merged could be merged as-is (when other issues have been addressed, etc.), so that I can work on top of a better master to implement exits
- doing a more direct implementation of exits (yet another parameter to `compile-low-expr` that we would pass all over the place) for now

I am going to "yield" this work now and move to other things, so in any case I will probably not write more code today.

Note: it is not as easy as I thought it would be to detect which actions are duplicated. The issue is that the PR currently mixes matrix decomposition and code generation (each decomposition of the matrix compute submatrices, generates their code, and then generates the code for the matrix by composing them under a `Switch`), so by the time the information comes back to the caller of how actions are used, the code is already generated. Two ideas:
1. I could give up on the idea of only using a duplication-avoidance mechanism only for actions that are indeed duplicated, and just use it for any "big enough" action.
2. I could reify the submatrix decomposition into a datastructure (where actions are represented by symbolic names / indices), get the duplication information from that, and then re-traverse it to generate the code.

I am thinking of going for (2), which is probably not that much work.

Reply:
Note: I did implement the first half of (2), now matrix-decomposition returns a "matching tree", which does not touch the action expressions themselves, and is lowered into code in a second step.